source/core/Variants.cpp

```diff
@@ -75,25 +87,42 @@ VIREO_FUNCTION_SIGNATURET(VariantToData, VariantToDataParamBlock)
         TypeRef inputType = _ParamImmediate(InputData._paramType);
         void* inputData = _ParamImmediate(InputData)._pData;
 
-        TypeRef destType = _ParamImmediate(DestData._paramType);
-        void* destData = _ParamImmediate(DestData)._pData;
+        TypeRef targetType = _ParamImmediate(TargetType._paramType);
+        void* targetData = _ParamImmediate(TargetType)._pData;
+
+        TypeRef outputType = _ParamImmediate(OutputData._paramType);
+        void* outputData = _ParamImmediate(OutputData)._pData;
 
         if (inputType->IsVariant()) {
             TypeRef variantInnerType = *reinterpret_cast<TypeRef *>_ParamImmediate(InputData._pData);
-            if (variantInnerType && variantInnerType->IsA(destType, true)) {
-                variantInnerType->CopyData(variantInnerType->Begin(kPARead), destData);
-            } else if (variantInnerType && destType->IsVariant()) {
-                *static_cast<TypeRef*>(destData) = CopyToVariant(variantInnerType);
+            if (variantInnerType
+                && IsTypeCompatibleWithInput(variantInnerType, targetType, targetData)
```

Comment:
1. targetData should not be used but only targetType should be used, because this corresponds to a type only input
2. It is safe to assume TargetType cannot be null and it is a required argument to this API. In NXG, this is going to be fixed to be a required terminal to match the CG behavior. So, in DFIR, if it is unwired, we can generate and pass a variant constant for now and assume here this will never be null.
3. targetType and outputType should be compared separately that they are identical. If they are not identical, we should bail out early. This is just an extra sanity check that DFIR already guarantees and we just want to ensure the same guarantee.
4. We should keep comparing variantInnerType against targetType separate, as we will be making it more liberal for coercible types. If we mix up these, then it seems to cause confusion we may end up with more restrictive or less restrictive than needed with 3-way type checking

Reply:
It is also probably fine to make the additional check (targetType and outputType) under a VIREO_ASSERT. As the test suite uses debug builds, it will catch any violations. This is a good safety net in Vireo for a very good guarantee we already receive from DFIR emitted VIA that ensure these two types are identical.

Reply:
1, 2. I am only passing in targetData to make it consistent with passing outputType and outputData. Having two varied APIs to compare input type with target vs output makes it harder to maintain. Additionally, it is possible for the targetType to be unavailable when it is unwired, and this can only be detected through targetData being null. (In this case, it means the targetType is a variant). While I can see how this can be avoided with better via emission, (which I in fact am) I think Vireo should be protected from bad inputs regardless. If we use the argument that "NXG doesn't allow this case", none of the changes in these new commits about outputType comparisons are necessary because SA protects us against them.
3. This is achieved by comparing target to input, and then output to input. (in the line below). I can add the additional VIREO_ASSERT that your next comment mentions.
4. Since what I have is correct for now, I would wait and make that change when we get to it. (I also think that the coercible type check will eventually apply for both the type comparisons.)


I vote to keep the comparison APIs simple. I say this because I did have multiple functions doing this at some point and it was hard for me to keep track even though I wrote them myself. If the complexity is necessary, we should do it whenever there's a need for it.

Reply:
Talked to Sankara IRL. Updated the code to:
1. check that the inputType is compatible with targetType (in the future, this function will verify that they are coercible)
2. check that the inputType is the same as the outputType (aka, they satisfy isA, or that the outputType is unwired, or both are unwired, or both are variants)